lib/phoenix/pubsub/local.ex

```diff
@@ -0,0 +1,212 @@
+defmodule Phoenix.PubSub.Local do
+  use GenServer
+
+  @moduledoc """
+  PubSub implementation for handling local-node process groups
+
+  This modules is used by Phoenix pubsub adapters to handle their
+  local node topic subscriptions. See `Phoenix.PubSub.PG2Adapter`
+  for an example integration.
+  """
+
+  @doc """
+  Starts the server
+
+    * `server_name` - The name to registered the server under
+
+  """
+  def start_link(server_name) do
+    GenServer.start_link(__MODULE__, [], name: server_name)
+  end
+  def start_link() do
+    GenServer.start_link(__MODULE__, [])
+  end
+
+
+  @doc """
+  Subscribes the pid to the topic
+
+    * `pid` - The subscriber Pid
+    * `topic` - The string topic, ie "users:123"
+
+  ## Examples
+
+      iex> PubSub.Local.subscribe(:local_server, self, "foo")
+      :ok
+
+  """
+  def subscribe(local_server, pid, topic) do
+    GenServer.call(local_server, {:subscribe, pid, topic})
+  end
+
+  @doc """
+  Unsubscribes the pid from the topic
+
+    * `pid` - The subscriber Pid
+    * `topic` - The string topic, ie "users:123"
+
+  ## Examples
+
+      iex> PubSub.Local.unsubscribe(:local_server, self, "foo")
+      :ok
+
+  """
+  def unsubscribe(local_server, pid, topic) do
+    GenServer.call(local_server, {:unsubscribe, pid, topic})
+  end
+
+  @doc """
+  Sends a message to allow subscribers of a topic
+
+    * `topic` - The string topic, ie "users:123"
+
+  ## Examples
+
+      iex> PubSub.Local.broadcast(:local_server, "foo")
+      :ok
+      iex> PubSub.Local.broadcast(:local_server, "bar")
+      :no_topic
+
+  """
+  def broadcast(local_server, topic, msg) do
+    GenServer.call(local_server, {:broadcast, :none, topic, msg})
+  end
+
+  @doc """
+  Returns the `HashSet` of subscribers pids for the given topic
+
+    * `topic` - The string topic, ie "users:123"
+
+  ## Examples
+
+      iex> PubSub.Local.subscribers(:local_server, "foo")
+      #HashSet<[]>
+
+  """
+  def subscribers(local_server, topic) do
+    GenServer.call(local_server, {:subscribers, topic})
+  end
+
+  @doc false
+  def list(local_server) do
+    GenServer.call(local_server, :list)
+  end
+
+  @doc false
+  def subscription(local_server, pid) do
+    GenServer.call(local_server, {:subscription, pid})
+  end
+
+  @doc false
+  def init(_) do
+    Process.flag(:trap_exit, true)
```

Comment:
We have discussed this locally but I will repeat it here.

We should likely `Process.flag(:trap_exit, true)` indeed. However, we need to add a test because the only reason we are adding this is because we want the subscriber to link to this process (effectively crashing the subscriber when the publisher dies). My proposal is to add a `:link` option to `subscribe` which automatically links both processes, providing a complete use case for this feature.